test-common/src/main/java/org/apache/kafka/common/test/TestUtils.java

```diff
@@ -125,4 +155,210 @@ static void retryOnExceptionWithTimeout(final Runnable runnable) throws Interrup
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                        return !checkpointFile.read().containsKey(tp);
+                    });
+                })),
+            "Cleaner offset for deleted partition should have been removed");
+
+        // Ensure that the topic directories are soft-deleted
+        TestUtils.waitForCondition(() -> brokers.stream().allMatch(broker ->
+                CollectionConverters.asJava(broker.config().logDirs()).stream().allMatch(logDir ->
+                    topicPartitions.stream().noneMatch(tp ->
+                        new File(logDir, tp.topic() + "-" + tp.partition()).exists()))),
+            "Failed to soft-delete the data to a delete directory");
+
+        // Ensure that the topic directories are hard-deleted
+        TestUtils.waitForCondition(() -> brokers.stream().allMatch(broker ->
+            CollectionConverters.asJava(broker.config().logDirs()).stream().allMatch(logDir ->
+                topicPartitions.stream().allMatch(tp ->
+                    Arrays.stream(Objects.requireNonNull(new File(logDir).list())).noneMatch(partitionDirectoryName ->
+                        partitionDirectoryName.startsWith(tp.topic() + "-" + tp.partition()) &&
+                            partitionDirectoryName.endsWith(UnifiedLog.DeleteDirSuffix())))
+            )
+        ), "Failed to hard-delete the delete directory");
+    }
+
+    public static int waitUntilLeaderIsElectedOrChangedWithAdmin(Admin admin,
+                                                                 String topic,
+                                                                 int partitionNumber,
+                                                                 long timeoutMs,
+                                                                 Optional<Integer> oldLeaderOpt,
+                                                                 Optional<Integer> newLeaderOpt) throws Exception {
+        if (oldLeaderOpt.isPresent() && newLeaderOpt.isPresent()) {
+            throw new IllegalArgumentException("Can't define both the old and the new leader");
+        }
+
+        GetPartitionLeader getPartitionLeader = (t, p) -> {
+            TopicDescription topicDescription = admin.describeTopics(Collections.singletonList(t)).allTopicNames().get().get(t);
+            return topicDescription.partitions().stream()
+                    .filter(partitionInfo -> partitionInfo.partition() == p)
+                    .findFirst()
+                    .map(partitionInfo -> partitionInfo.leader().id() == Node.noNode().id() ? null : partitionInfo.leader().id());
+        };
+
+        return doWaitUntilLeaderIsElectedOrChanged(getPartitionLeader, topic, partitionNumber, timeoutMs, oldLeaderOpt, newLeaderOpt);
+    }
+
+    private static int doWaitUntilLeaderIsElectedOrChanged(GetPartitionLeader getPartitionLeader,
```

Comment:
Could you please try to simply `waitUntilLeaderIsElectedOrChangedWithAdmin` and `doWaitUntilLeaderIsElectedOrChanged`?